content/Building an Engine-modularized API in Rails - Guillermo Leopold.md

```diff
@@ -0,0 +1,154 @@
+    * It might be useful to create a few scripts to run migrations, run rspec or bundle install across all the engines. The scripts will iterate over the engines folder and run the command inside each one.
+
+On the other hand, testing inside an engine with a rspec it’s quite similar to what we are used to doing and it provides us with some guarantees and the only considerations that took us some time are:
+  * We will need to load manually factories from other engines when using FactoryBot in our engine
+  * We need to require in our rails_helper.rb any development dependency that we are using in our test suite
+
+After developing a feature, one of the advantages that we found out about working with engines is that we can just run the test suite for the engine where we have been working on without the necessity to go through all the suite tests in the application. This reduces the time to run tests locally dramatically.
+
+The API architecture and the project structure will look something like this:
+
+![Engines Structure + Folder Struct](images/engines_structure_folder_struct.png)
+
+## Summing up
+
+It’s been a while since we started working with this project structure. Right now we are at a point where we can already look back and find the pros and cons of making this particular decision:
+
+**Pros:**
+  * Helps to parallelize work when working with medium-sized development teams.
+  * Makes the code more organized and easier to read or find where a feature should be implemented.
+  * The possibility to run tests inside an engine scope reduces time when testing locally.
+  * A modular application like this is a mid-step between a monolith API and microservices.
+  * Possibility to pick a full engine and use it in another API. We did this with our catalog engine and an admin that we develop in another project.
+
+
+**Cons:**
+  * Lot of configuration and changes are needed when starting working with engines, requires a bit of time during the first weeks.
+  * Lack of information on the internet, building an API with engines doesn’t seem to be a popular choice.
+  * Despite our effort to avoid repeating code, we still repeat more code than when working in a monolith api
+
+After overcoming the initial learning curve, we feel quite comfortable working with engines and we found out the benefits of having chosen this architecture instead of a monolith one, so I think that I would use it again if it fits the project requirements. In spite of that, we still have a lot to learn and improve upon what we already did.
+
+As a pending feature, we would also like to see how we could take advantage of this architecture and try to deploy our application into microservices; of course that implies some changes will need to be made but it would be a valuable achievement.
+
+## Learning about engines
+
+To learn more about engines, the [Ruby on Rails official documentation](https://guides.rubyonrails.org/engines.html) is a good follow-on reading.
+
```

Comment:
can you also add a reference to that medium post in which we based (I think anthony recommended it)